### ISSUE TITLE
Assert during dependency graph construction on conan install

### ISSUE THREAD

**Issue by HC-pel** [conan-io/conan#5685]
Conan crashes during the calculation of the dependency graph under certain circumstances. I've uploaded a minimally reproducing example here: https://github.com/HC-pel/conan_assert
Removing any of the (normal or build) requirements in any of the conanfiles prevents the assert. The assert also does not trigger if there is a package available for build_req, either locally or in a remote. The backtrace looks as follows:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/conans/client/command.py", line 1859, in run
    method(args[0][1:])
  File "/usr/local/lib/python2.7/dist-packages/conans/client/command.py", line 471, in install
    lockfile=args.lockfile)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/conan_api.py", line 78, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/conan_api.py", line 577, in install
    no_imports=no_imports)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/manager.py", line 60, in install
    self._recorder)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_manager.py", line 182, in load_graph
    graph_lock=graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_manager.py", line 284, in _load_graph
    apply_build_requires=apply_build_requires)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_manager.py", line 255, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_manager.py", line 251, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_builder.py", line 65, in extend_build_requires
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_builder.py", line 178, in _handle_require
    update, remotes, processed_profile, graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_builder.py", line 132, in _load_deps
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph_builder.py", line 200, in _handle_require
    dep_graph.add_edge(node, previous, require)
  File "/usr/local/lib/python2.7/dist-packages/conans/client/graph/graph.py", line 213, in add_edge
    assert src in self.nodes and dst in self.nodes
AssertionError
```

The operating system or Python version does not seem to have any effect, I've tested it in various combinations of Ubuntu, Windows, Python 2 and Python 3. It still happens in master as of the time of this writing.

This is potentially a duplicate of #5612, however unlike that issue I am not using workspaces. The first major version to have this issue is Conan 1.10, it does not happen with 1.9 or earlier.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
I have managed to reproduce the assertion, thanks @HC-pel for providing a reproducible case!
Seems a difficult issue with multiple nesting of build-requires using regular requires that are in turn shared with regular requires in the "main" graph. I am investigating possible mitigations.

**Comment by memsharded** [comment]
This has been fixed in #5689 and will be released in 1.18.2 today. Thanks again for reporting it.
